### ISSUE TITLE
takeUntil does not stop in rxjs 6

### ISSUE THREAD

**Issue by dxcqcv** [ReactiveX/rxjs#4277]
## Bug Report

**Current Behavior**
TLDR: To clear the input, then the submit still log.

long:  have 2 inputs and a submit button, and console.log submit when clicked button and 2 inputs be verified, and if one of inputs failed then button do nothing.

I almost finish this feature, but have a weird problem button log when clicked and 2 inputs passed, but when one input failed again, the button still log, and only two inputs failed then it stop, I don't know why, I wanna one of inputs failed then button stop, how do I to fix it? thx

**Reproduction**
- REPL or Repo link:
https://codepen.io/dxcqcv/pen/BqVGJP?editors=0010

```js
   class Main extends React.Component {
    constructor(props) {
    super(props);

    this.phoneNum = "";

    this.state = {
      picCodeVal: ""
    };

    this.picCodeSubject$ = new rxjs.Subject();
    this.phoneNumSubject$ = new rxjs.Subject();

    this.phoneDom = React.createRef();
    this.registerDom = React.createRef();

    this.handlePicCodeChange = this.handlePicCodeChange.bind(this);
  }
  componentDidMount() {
    this.verifyPhone();
    this.registerUser();
  }

  verifyPhone() {
    const [phoneNum$, phoneStr$] = rxjs.fromEvent(
      this.phoneDom.current,
      "input"
    ).pipe(
                  rxjs.operators.map(x =>"1"),// codepen cannot get input value with data so I change to this
  //map(val => val.data),


      rxjs.operators.scan((acc, cur) => (cur ? [...acc, ...cur] :    acc.slice(0, -1)), []),

      rxjs.operators.partition(arr => arr.every(isNumber))
    );
    const [phoneNumErr$, verifiedPhoneNum$] = phoneNum$.pipe(

      rxjs.operators.map(arr => arr.length),
      rxjs.operators.partition(l => l === 0 || l < 11)
    );
    const failedPhoneNum$ = rxjs.merge(phoneStr$, phoneNumErr$);

    failedPhoneNum$.subscribe(x => {
      this.phoneNum = "";
      this.phoneNumSubject$.next(this.phoneNum);
    });
    verifiedPhoneNum$.subscribe(x => {
      // x is 11
      console.log("verified phone",x)
      this.phoneNum = x;
      this.phoneNumSubject$.next(this.phoneNum);
    });
  }

  handlePicCodeChange({ target: { value } }) {
    this.setState({ picCodeVal: value });
    this.picCodeSubject$.next(value);
  }
  registerUser() {
    const clickReg$ = rxjs.fromEvent(this.registerDom.current, "click");

        const [stopReg$, right$] = rxjs.combineLatest(
      this.picCodeSubject$,
      this.phoneNumSubject$
    ).pipe(rxjs.operators.partition(arr => arr.some(isEmpty)));
    const restoreReg$ = right$.pipe(
      rxjs.operators.take(1),
      rxjs.operators.repeatWhen(() => stopReg$)
    );

        clickReg$
      .pipe(

        rxjs.operators.takeUntil(stopReg$),
        rxjs.operators.repeatWhen(() => restoreReg$)
      )
      .subscribe(x => console.log("submit"));

    stopReg$.subscribe(x => console.log("stop stream", x));
    restoreReg$.subscribe(x => console.log("restore stream", x));

    // init phoneNum and picCode vals
    this.phoneNumSubject$.next(this.phoneNum);
    this.picCodeSubject$.next(this.state.picCodeVal);


  }
  render() {
    const { picCodeVal } = this.state;
    return (
      <div>
        <p>
          <input placeholder="phone" maxLength="11" ref={this.phoneDom} />
        </p>
        <p>
          <input
            placeholder="pic code"
            maxLength="11"
            value={picCodeVal}
            onChange={this.handlePicCodeChange}
          />
        </p>

        <button styleName="register-btn" ref={this.registerDom}>
          submit
        </button>
      </div>
    );
  }
}

ReactDOM.render(<Main />, document.getElementById("root"));

function isNumber(n) {
  return !isNaN(parseFloat(n)) && isFinite(n);
}

function isEmpty(n) {
  return n === "";
}
```

**Expected behavior**
When one of inputs fail to verify, then the submit button do nothing, when two inputs all pass the verification, then the button will be restored

**Environment**
- Runtime:  Chrome 67
- RxJS version: 6.3.3
- react 16.5.2

**Possible Solution**
I do not know

**Additional context/Screenshots**
I also ask in https://stackoverflow.com/questions/52912709/takeuntil-does-not-stop-in-rxjs-6, but no one answer it

**Comment by cartant** [comment]
This seems more like a why-doesn't-my-code-work question than a bug report.

Could you please try to make a minimal reproduction of precisely what it is that you consider to be a bug?

**Comment by dxcqcv** [comment]
@cartant @kwonoj hi, I updated look it if you have time, thx

**Comment by kqadem** [comment]
Okay, I'm here to proof you coded completely ******. I just reordered the positions and renamed to understand it

`https://codepen.io/anon/pen/BqMEvd?editors=0010`
As you can see:

- you emit origin1 and origin2. (73 & 74)
- after that, the combinedLatest triggers first time and logs "origins emitted" twice. (30)
- by now, empty$ = ["", ""] and notEmpty$ = []
- you subscribe to empty$ (66) and receive the ["",""] + the log
- and since empty$ has already emitted, start$ won't work because completed
- after you click both restore1 and restore2 button, your completely stupid logic on line 29 will result in - - empty$=[] and notEmpty$=["Restored 1, Restored 2"]
- takeUntil won't be triggered anymore (57)
- but start$ is repeated because restoreReg emitted and WORKS NOW (58)

**Comment by kqadem** [comment]
Sry for my phrasing, but reading 3 times about this guy the same day made me crazy. Here + https://www.reddit.com/r/rxjs/comments/9q2rqw/takeuntil_does_not_stop_in_rxjs_6/ +
https://stackoverflow.com/questions/52912709/weird-takeuntil-behave-in-rxjs-6/53018241#53018241

**Comment by kqadem** [comment]
Saw it coming that way. I'd still be okay if you had deleted the second post also. Despite I'm still amazed about the reactive paradigm & rxjs after 2 years+, this Perfectly illustrates why I'm not contributing. I couldn't stand the way how the majority treats the open source devs. Kudos on you dealing with them

**Comment by dxcqcv** [comment]
First, the start button is stopped, then click restore01 button and restore02 button the start button restored,

then the bug is when click stop button01 or stop button02 then the stop button should be stopped, but not

notice console.log, the start button is stopped when start, then you click restore01 and restore02, then the start button works, and now the bug is coming like winter, when I click stop01 button or stop02 button, in console.log it show stop stream Array [ "r", "" ] but the start button still works, not be stopped, can you image my feeling about see this circumstances? oh my god

[online demo](https://codepen.io/dxcqcv/pen/QZBEqJ?editors=1010)

the code is blow

    const startDom = document.querySelector("#start")
    const numberDom = document.querySelector("#number")
    const restoreDom01 = document.querySelector("#restore01")
    const stopDom01 = document.querySelector("#stop01")
    const restoreDom02 = document.querySelector("#restore02")
    const stopDom02 = document.querySelector("#stop02")
    
    let init01 = false
    let init02 = false
    
    const origin01$ = new rxjs.Subject()
    const origin02$ = new rxjs.Subject()
    
    const restore01$ = rxjs.fromEvent(restoreDom01, "click")
    restore01$.subscribe(() => origin01$.next("r"))
    const stop01$ = rxjs.fromEvent(stopDom01, "click")
    stop01$.subscribe(() => origin01$.next(""))
    
    const restore02$ = rxjs.fromEvent(restoreDom02, "click")
    restore02$.subscribe(() => origin02$.next("r"))
    const stop02$ = rxjs.fromEvent(stopDom02, "click")
    stop02$.subscribe(() => origin02$.next(""))
    
    const [stopReg$, right$] = rxjs
    .combineLatest(origin01$, origin02$)
    .pipe(rxjs.operators.partition(arr => arr.some(isEmpty)));
    const restoreReg$ = right$.pipe(
      rxjs.operators.take(1),
      rxjs.operators.repeatWhen(() => stopReg$)
    );
    
    
    const start$ = rxjs.fromEvent(startDom, "click");
    start$
      .pipe(
      rxjs.operators.mapTo(1),
      rxjs.operators.scan((acc, cur) => acc + cur, 0),
      rxjs.operators.takeUntil(stopReg$),
      rxjs.operators.repeatWhen(() => restoreReg$)
    )
      .subscribe(x => (numberDom.innerHTML = x));
    
    
    stopReg$.subscribe(x => console.log("stop stream", x));
    restoreReg$.subscribe(x => console.log("restore stream", x));
    
    origin01$.next("")
    origin02$.next("")
    
    function isEmpty(n) {
      return n === "";
    }


may be it's not very clear, so I tell you the steps

1.  click 2 restore button, one by one

2.  click start button, you'll see the number be added one after clicked

3.  click stop button, OK, this is a problem, for now start button should be stopped, but not, you still can add one to click start button, but in console you can see the stop stream is subscribed, so weird

[I also have other demo with the same problem using react and rxjs](https://codepen.io/dxcqcv/pen/BqVGJP?editors=0010)

**Comment by cartant** [comment]
For this issue to be treated as a bug report, it needs both a simpler reproduction and step-by-step instructions that are clear.

Ideally, if you believe that you have found a bug in the implementation of `takeUntil` - which is what this issue's title claims - you should be able to provide a failing test. However, a simple reproduction will suffice.

ATM, your example is too complicated and the behaviour you are expecting is unclear.

**Comment by kqadem** [comment]
> > ... [The problem does not appear to be with the implementation of `takeUntil`] ...
> > I just reordered the positions and renamed to understand it
> > https://codepen.io/anon/pen/BqMEvd?editors=0010
> > As you can see:
> > ```
> > * you emit origin1 and origin2. (73 & 74)
> > 
> > * after that, the combinedLatest triggers first time and logs "origins emitted" twice. (30)
> > 
> > * by now, empty$ = ["", ""] and notEmpty$ = []
> > 
> > * you subscribe to empty$ (66) and receive the ["",""] + the log
> > 
> > * and since empty$ has already emitted, start$ won't work because completed
> > 
> > * after you click both restore1 and restore2 button, ... [the] ... logic on line 29 will result in empty$=[] and notEmpty$=["Restored 1, Restored 2"]
> > 
> > * takeUntil won't be triggered anymore (57)
> > 
> > * but start$ is repeated because restoreReg emitted and WORKS NOW (58)
> > ```
> 
> In my opinion you cannot understand what about my question, your answer is only I expected and worked because this code is I made, my question is after start$ be repeated, then click stop button, start$ should be stopped, but not, why ? can you explain that ?

Interesting how you assume my ability of understanding. Are you not the one in the need of help? So you are in the position to provide clear examples in order that others can help you, don't you think?
Anyway, to cut it down:
- you click stop
   1. start$ is stopped because takeUntil (38)
   2. restoreReg is repeated (29)
   3. start$ is repeated (39)

start$ is not stopped, because after it is repeated there is no emit of stopReg$, so takeUntil does not trigger.
This time the lines are used [on your example](https://codepen.io/dxcqcv/pen/QZBEqJ?editors=1010)


But as @cartant already said, if you really think there is bug, provide precise example. Maybe there is a race condition in your real code. Since I think this is nothing rxjs related anymore, I'm done here. And sry again for my phrasing @cartant

**Comment by dxcqcv** [comment]
sorry for later, if as you say when I click stop, then `restoreReg is repeated (29)` is be triggered, why by this time only log `stop stream Array [ "", "r" ]`

and if I commented line 29, start$ is still not stopped, so your guess is wrong, and no a race condition, because 100% cannot stop

I also need time to provide precise example, sorry

**Comment by kwonoj** [comment]
I'm going to close this as issue has opened some time period without discrete reproducible code, while most of behavior was answered in current code's behavior not related to core library's bug.

If you still feel there is a bug in rxjs, feel free to create new issue with

- reproducible (node.js env is preferred, but either's fine) does not related with user logics
- elaborated details for reason of it's core bug, instead of user level code issue.
